### DISCUSSION TITLE
Fix random seed for permutation

### DISCUSSION THREAD

**Discussion by yiran25** [bonsai-rx/bonsai#1354] [Q&A]
Hi I was wondering if I can fix the sequence for permutation via some random seed?

Basically I have 3 parameters. Parameter A, B, C. Parameter A is chosen randomly from a list of 3 integers ([1,2,3]). Parameter B is always ordered ([10,20,30]) and is created by a `ParameterRange`. Parameter C is randomly chosen from a `CreateUniformDistribution`. 

In the end, each trial will run through all the 3 possible combinations of B and C (each B gets paired with a random C, so for each trial, there will be B=10, C=x; B=20, C=y; B=30, C=z). And each of the 3 combinations of B&C will be paired with a single A value randomly chosen from the list. So for each round, the possible result will be like the following: 
A=2, B=10, C=x1
A=2, B=20, C=x2
A=2, B=30, C=x3
A=1, B=10, C=y1
A=1, B=20, C=y2
A=1, B=30, C=y3
A=3, B=10, C=z1
A=3, B=20, C=z2
A=3, B=30, C=z3
And then the next round of 9 possibilities will continue, with A randomised in a different order (say 1,2,3 this time). and so on. 

What I want to achieve is that after I stop the workflow and start it again, I have exact the same sequence of A,B, and C as I get from the previous run of the workflow. so A will following the pattern of (2,1,3,1,2,3...) and C will have the same sequence of values (x1,x2,x3,y1,y2,y3...). How may I achieve this?

The workflow I'm using is here:
[test.zip](https://github.com/bonsai-rx/bonsai/files/11444590/test.zip)

A, B and C are chained up by `ParameterRange - CreateObservable - Concat`
![image](https://github.com/bonsai-rx/bonsai/assets/74373973/e8a1ca3e-6a2c-44ff-8796-4f03e7692d06)
A is just a `ParameterRange` with a permutation.
![image](https://github.com/bonsai-rx/bonsai/assets/74373973/de903cc3-bafe-4587-887d-20f5d6d81eac)
B and C are chained up by the same `ParameterRange - CreateObservable - Concat`logic.
![image](https://github.com/bonsai-rx/bonsai/assets/74373973/eff5eaa9-a88c-4539-bd64-d2dfcbf29893)
and B is just another parameter range.
![image](https://github.com/bonsai-rx/bonsai/assets/74373973/22147c1b-dcee-48b5-bfd8-9e15e5f26659)
C is a `CreateUniformDistribution` - `Permutation` - `Sample`
![image](https://github.com/bonsai-rx/bonsai/assets/74373973/b512af52-e0d4-4c3d-ad1e-7a973655724c)
If I fix the seed for the `CreateRandom` C will always just be one number, which is not ideal.
Many thanks!

**Comment by bruno-f-cruz** [comment]
I am not sure I completely followed, but I think the pattern you are looking for is something like:

![image](https://github.com/bonsai-rx/bonsai/assets/7049351/f1281a97-97c2-4161-ae45-d4c73620063a)

At the start of the workflow, you save a `CreateRandom` object that will be used to generate the pseudo-random values later on.
You then use this object as an input to `CreateContinuousUniform` to ensure that the same values will always be drawn. Next, you Sample as many values as you need (`SampleMany`) and `Permutation` to shuffle them. Now, the trick to always get the same order after the permutation is to also input the `CreateRandom` object to the `Permutation` node, leading to a deterministic behavior on each shuffle sequence.

Hope it helps!


<details>
    <summary>Workflow (Copy+Paste me)</summary>
    
```xml
<?xml version="1.0" encoding="utf-8"?>
<WorkflowBuilder Version="2.7.3"
                 xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                 xmlns:rx="clr-namespace:Bonsai.Reactive;assembly=Bonsai.Core"
                 xmlns:p1="clr-namespace:Bonsai.Numerics.Distributions;assembly=Bonsai.Numerics"
                 xmlns:num="clr-namespace:Bonsai.Numerics;assembly=Bonsai.Numerics"
                 xmlns="https://bonsai-rx.org/2018/workflow">
  <Workflow>
    <Nodes>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="IntProperty">
          <Value>123123123</Value>
        </Combinator>
      </Expression>
      <Expression xsi:type="PropertyMapping">
        <PropertyMappings>
          <Property Name="Seed" />
        </PropertyMappings>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="num:CreateRandom">
          <num:Seed>123123123</num:Seed>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="rx:Take">
          <rx:Count>1</rx:Count>
        </Combinator>
      </Expression>
      <Expression xsi:type="rx:AsyncSubject">
        <Name>Seed</Name>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>Seed</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="p1:CreateContinuousUniform">
          <p1:Lower>0</p1:Lower>
          <p1:Upper>1</p1:Upper>
        </Combinator>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="p1:SampleMany">
          <p1:Count>4</p1:Count>
        </Combinator>
      </Expression>
      <Expression xsi:type="SubscribeSubject">
        <Name>Seed</Name>
      </Expression>
      <Expression xsi:type="Combinator">
        <Combinator xsi:type="num:Permutation" />
      </Expression>
    </Nodes>
    <Edges>
      <Edge From="0" To="1" Label="Source1" />
      <Edge From="1" To="2" Label="Source1" />
      <Edge From="2" To="3" Label="Source1" />
      <Edge From="3" To="4" Label="Source1" />
      <Edge From="5" To="6" Label="Source1" />
      <Edge From="6" To="7" Label="Source1" />
      <Edge From="7" To="9" Label="Source1" />
      <Edge From="8" To="9" Label="Source2" />
    </Edges>
  </Workflow>
</WorkflowBuilder>

```

</details>
